### DISCUSSION TITLE
遇到一个替换了 IP 就没法正常工作的情况

### DISCUSSION THREAD

**Discussion by Journalist-HK** [XIU2/CloudflareSpeedTest#411] [提问] [answered]
有一个游戏[雀魂](https://game.mahjongsoul.com/)，两个域名 mjjpgs.mahjongsoul.com passport.mahjongsoul.com 用的是CF，但是只要换了 IP 就出错。（mahjongsoul.com 这个域名有 DNS 污染，我说的是已经解决了 DNS 污染的情况）

**Comment by charSLee013** [comment]
我试了一下用`1.1.1.1`解析得到的IP地址进行访问

```bash
➜  ~ dig @1.1.1.1 passport.mahjongsoul.com +short|nali
passport.mahjongsoul.com.cdn.cloudflare.net [Cloudflare] .
8fd5a82d380649ccbe0b48daa2abc136.pacloudflare.com.
172.65.229.49 [美国 CloudFlare节点]

# 指定IP地址和端口访问
➜  ~ curl --resolve passport.mahjongsoul.com:443:172.65.229.49 https://passport.mahjongsoul.com/cdn-cgi/trace

h=passport.mahjongsoul.com
ts=1690014734.016
visit_scheme=https
uag=curl/7.68.0
colo=SIN
```
看起来没有问题而且`colo`是 SIN

换一个不同地区的IP地址进行访问
```bash
➜  ~ curl --resolve passport.mahjongsoul.com:443:198.41.194.98 https://passport.mahjongsoul.com/cdn-cgi/trace

h=passport.mahjongsoul.com
ts=1690014832.269
visit_scheme=https
uag=curl/7.68.0
colo=NRT
```

看起来能正常连接没问题...

**Reply by Journalist-HK** [reply]
实际情况要复杂一点。先看 passport.mahjongsoul.com，对于上面的web游戏，没有问题，但是实际上在iOS上的客户端会连接 passport.mahjongsoul.com:9092
```
curl --resolve passport.mahjongsoul.com:9092:172.65.229.49 https://passport.mahjongsoul.com:9092/cdn-cgi/trace

h=passport.mahjongsoul.com
ts=1690021823.946
visit_scheme=https
uag=curl/8.1.2
colo=SJC
sliver=none
http=http/2
tls=TLSv1.2
sni=plaintext
warp=off
gateway=off
rbi=off
kex=X25519
```
换个 IP 就不行。

类似地，游戏会向 mjjpgs.mahjongsoul.com:8443 请求一个服务器地址（mjjpgs.mahjongsoul.com:9663）然后建立 websocket 连接。

![image](https://github.com/XIU2/CloudflareSpeedTest/assets/50702502/9b24a115-bc17-43d8-a0c1-8e0b158197af)

![image](https://github.com/XIU2/CloudflareSpeedTest/assets/50702502/f8a88a50-44cc-4371-ab96-783c3715150c)


一但换一个IP，mjjpgs.mahjongsoul.com:8443 没问题，mjjpgs.mahjongsoul.com:9663 会连接失败

![image](https://github.com/XIU2/CloudflareSpeedTest/assets/50702502/def41bf9-d3ae-4ab0-b19f-186cfe912c12)

**Reply by XIU2** [reply, marked as answer]
普通节点可能只开放了[标准 HTTP 端口](https://developers.cloudflare.com/fundamentals/get-started/reference/network-ports/)（例如 80 8080 443 8443 这些），像这种游戏服务器，走的端口都比较特殊，而且协议什么的也可能会比较特殊，因此只能使用分配的专用节点（IP）。

你截图里的信息，显示的就是该 IP 的 9663 端口超时，我测了下该端口压根就没有开放，自然无法使用啦。
估计只有默认分配的专用节点 IP 才开放了 9663 端口。

这种情况在游戏服务器上很常见，我以前也遇到过一个游戏服务器是 Cloudflare CDN，也和你这个一样，只能指向默认分配的节点。

**Comment by mengguyi** [comment]
这个IP可能是专门分给雀魂的，是不能随便指定的。就像discord

**Comment by charSLee013** [comment]
你可以试试 **47.252.21.16** **47.89.136.64** 这两个IP替换掉`mjjpgs.mahjongsoul.com:9663`是否可以访问成功 👀

**Reply by Journalist-HK** [reply]
可以，都是雀魂的服务器，证书是对的。mjusgsbk.mahjongsoul.com mjusgs.mahjongsoul.com
